docs/sql-migration-guide-upgrade.md

```diff
@@ -147,6 +147,8 @@ license: |
 
   - Since Spark 3.0, if files or subdirectories disappear during recursive directory listing (i.e. they appear in an intermediate listing but then cannot be read or listed during later phases of the recursive directory listing, due to either concurrent file deletions or object store consistency issues) then the listing will fail with an exception unless `spark.sql.files.ignoreMissingFiles` is `true` (default `false`). In previous versions, these missing files or subdirectories would be ignored. Note that this change of behavior only applies during initial table file listing (or during `REFRESH TABLE`), not during query execution: the net change is that `spark.sql.files.ignoreMissingFiles` is now obeyed during table file listing / query planning, not only at query execution time.
 
+  - Since Spark 3.0, substitution order of nested WITH clauses is changed and an inner CTE definition takes precedence over an outer. The previous behaviour can be restored by setting `spark.sql.legacy.cte.substitution.enabled` to `true`.
```

Comment:
Ur, IIUC, `nested WITH clause` is a new one in Spark 3.0.0, isn't it, @peter-toth ?

Reply:
https://github.com/apache/spark/pull/24831 added support for `WITH` clause in subqueries and introduced nesting like here:
```
WITH t(c) AS (SELECT 1)
SELECT max(c) FROM (
  WITH t(c) AS (SELECT 2)
  SELECT * FROM t
)
```

But other kinds of nested `WITH` clauses were available even before that PR like:
- Nesting in CTE:
  ```
  WITH
    t AS (SELECT 1),
    t2 AS (
      WITH t AS (SELECT 2)
      SELECT * FROM t
    )
  SELECT * FROM t2
  ```
- Nesting in subquery expression:
  ```
  WITH t AS (SELECT 1)
  SELECT (
    WITH t AS (SELECT 2)
    SELECT * FROM t
  )
  ```
The issue this PR fixes is that the last 2 examples should return `2` instead of `1`.

Reply:
Thanks.